### PR TITLE
Bug #2059: Apply +pemap and +commap in terms of hwloc PUs

### PR DESCRIPTION
**Pull request by evan-charmworks**
I believe this should help with bug #2059, but I will need someone with access to and more experience with various machine layers to confirm that the issue is resolved.

### CODE REVIEW

**Comment by nitbhat** [comment]
I'll test this out tomorrow.

File: src/conv-core/cpuaffinity.C [hunk 1]
```diff
@@ -204,8 +204,12 @@ int CmiSetCPUAffinity(int mycore)
...
   cmi_hwloc_topology_load(topology);
 
-  hwloc_cpuset_t cpuset = cmi_hwloc_bitmap_alloc();
-  cmi_hwloc_bitmap_set(cpuset, core);
+  int thread_unitcount = cmi_hwloc_get_nbobjs_by_type(topology, HWLOC_OBJ_PU);
+  int thread_assignment = core % thread_unitcount;
+
+  hwloc_obj_t thread_obj = cmi_hwloc_get_obj_by_type(topology, HWLOC_OBJ_PU, thread_assignment);
+  hwloc_cpuset_t cpuset = cmi_hwloc_bitmap_dup(thread_obj->cpuset);
+  cmi_hwloc_bitmap_singlify(cpuset);
```
Comment (rbuch) [inline src/conv-core/cpuaffinity.C]: Can you add some comments to this saying what the hwloc functions are each doing? I, for example, am unclear on why singlify needs to be called at all. Shouldn't there just be one bit set in the first place?

**Comment by jcphill** [comment]
What does this change from a user-visible perspective?

**Comment by evan-charmworks** [comment]
Before this change, the implementation of pemap using hwloc as a substrate would apply the specified values directly to what hwloc calls a bitmap. I suspect this to be in error, and it seems to sometimes cause `Invalid argument` and not actually bind things, so this change adjusts it to instead generate a bitmap through the proper hwloc API, in this case by treating the specified values as index numbers of Processing Units (i.e. SMTs or hardware threads, the smallest unit of CPU).

**Comment by jcphill** [comment]
There is a good description at https://www.open-mpi.org/projects/hwloc/doc/v2.0.4/a00326.php#faq_indexes and https://www.open-mpi.org/projects/hwloc/doc/v2.0.4/a00313.php of physical/OS vs logical indices. Since +pemap and +commap have traditionally used physical indices, it would be extremely disruptive to change their meaning to logical indices, especially on Intel platforms. If logical indices are desired they should either use new options or require a new flag.

**Comment by evan-charmworks** [comment]
> Don't break existing +pemap/+commap behavior.

@jcphill I've updated this PR to interpret these parameters as physical IDs. Does this allow `+pemap 0-83:4,88-171:4` to work as intended?

**Comment by jcphill** [comment]
> > Don't break existing +pemap/+commap behavior.
> 
> @jcphill I've updated this PR to interpret these parameters as physical IDs. Does this allow `+pemap 0-83:4,88-171:4` to work as intended?

You should be able to test this on any hyperthreaded Intel machine, as the logical and physical ordering will be different.

**Comment by evan-charmworks** [comment]
> > > Don't break existing +pemap/+commap behavior.
> > 
> > 
> > @jcphill I've updated this PR to interpret these parameters as physical IDs. Does this allow `+pemap 0-83:4,88-171:4` to work as intended?
> 
> You should be able to test this on any hyperthreaded Intel machine, as the logical and physical ordering will be different.

I don't understand what you mean.

**Comment by jcphill** [comment]
> > > > Don't break existing +pemap/+commap behavior.
> > > 
> > > 
> > > @jcphill I've updated this PR to interpret these parameters as physical IDs. Does this allow `+pemap 0-83:4,88-171:4` to work as intended?
> > 
> > 
> > You should be able to test this on any hyperthreaded Intel machine, as the logical and physical ordering will be different.
> 
> I don't understand what you mean.
Can you confirm that +pemap works as before on a hyperthreaded Intel machine, because I expect that the original version of the patch would have altered it.

**Comment by jcphill** [comment]
On Summit, the installed lstopo seems to give accurate results. Is there a way to have Charm++ link against the system hwloc library rather than the internal one?

**Comment by evan-charmworks** [comment]
What is the output of `lstopo --version` on Summit? I doubt there are any local changes in the installed hwloc that have not been upstreamed. If Charm is still reporting incorrect values then I would bet there is more work to be done in our code.

**Comment by jcphill** [comment]
lstopo 1.11.8

**Comment by evan-charmworks** [comment]
We currently embed 1.11.12.

**Comment by evan-charmworks** [comment]
On a Summit job where resource limiting is in effect (whatever circumstances lead Charm++ to report less than the whole system as present), what are the outputs of these commands?

`lstopo --only PU`

`lstopo --only PU --whole-system`

**Comment by jcphill** [comment]
I've looked at jsrun lstopo [--whole-system] outputs (without --only PU) and they look correct for the system-installed lstopo.

**Comment by jcphill** [comment]
OK, I see what's going on now. There is a PBS "-alloc_flags smt1" option that switches from the default smt4 mode to smp1 at some deeper level in the OS. In smt4 you see all of the PUs, but in smt1 you get this:
--bind rs -p1 -r6 -g1 -c7
PU L#0 (P#0)
PU L#1 (P#4)
PU L#2 (P#8)
PU L#3 (P#12)
PU L#4 (P#16)
PU L#5 (P#20)
PU L#6 (P#24)
--bind rs -p1 -r6 -g1 -c7 --whole-system
PU L#0 (P#0)
PU L#1 (P#4)
PU L#2 (P#8)
PU L#3 (P#12)
PU L#4 (P#16)
PU L#5 (P#20)
PU L#6 (P#24)
PU L#7 (P#28)
PU L#8 (P#32)
PU L#9 (P#36)
PU L#10 (P#40)
PU L#11 (P#44)
PU L#12 (P#48)
PU L#13 (P#52)
PU L#14 (P#56)
PU L#15 (P#60)
PU L#16 (P#64)
PU L#17 (P#68)
PU L#18 (P#72)
PU L#19 (P#76)
PU L#20 (P#80)
PU L#21 (P#84)
PU L#22 (P#85)
PU L#23 (P#86)
PU L#24 (P#87)
PU L#25 (P#88)
PU L#26 (P#92)
PU L#27 (P#96)
PU L#28 (P#100)
PU L#29 (P#104)
PU L#30 (P#108)
PU L#31 (P#112)
PU L#32 (P#116)
PU L#33 (P#120)
PU L#34 (P#124)
PU L#35 (P#128)
PU L#36 (P#132)
PU L#37 (P#136)
PU L#38 (P#140)
PU L#39 (P#144)
PU L#40 (P#148)
PU L#41 (P#152)
PU L#42 (P#156)
PU L#43 (P#160)
PU L#44 (P#164)
PU L#45 (P#168)
PU L#46 (P#172)
PU L#47 (P#173)
PU L#48 (P#174)
PU L#49 (P#175)

**Comment by jcphill** [comment]
So the assumption that the OS PU indices are 0...ncpus-1 is incorrect.

**Comment by evan-charmworks** [comment]
@jcphill I pushed a commit that removes the `Error> Invalid core number` check entirely, since we should be able to rely on hwloc's return code to tell if there actually was a problem. I'm wondering now if 1) the affinity issue is fixed 2) whether `Warning: the number of SMP threads` still prints.

File: src/conv-core/cpuaffinity.C [hunk 2]
```diff
@@ -932,10 +932,6 @@ void CmiInitCPUAffinity(char **argv)
   if (pemap != NULL && CmiMyPe()<CmiNumPes()) {    /* work thread */
     int mycore = search_pemap(pemap, CmiMyPeGlobal());
     if(show_affinity_flag) CmiPrintf("Charm++> set PE %d on node %d to core #%d\n", CmiMyPe(), CmiMyNode(), mycore);
-    if (mycore >= CmiNumCores()) {
-      CmiPrintf("Error> Invalid core number %d, only have %d cores (0-%d) on the node. \n", mycore, CmiNumCores(), CmiNumCores()-1);
-      CmiAbort("Invalid core number");
-    }
     if (CmiSetCPUAffinity(mycore) == -1) CmiAbort("set_cpu_affinity abort!");
```
Comment (jcphill) [inline src/conv-core/cpuaffinity.C]: Can you add more information to this error message (e.g., "Error setting PE %d on node %d to core #%d")?

Comment (evan-charmworks) [reply src/conv-core/cpuaffinity.C]: Done

**Comment by jcphill** [comment]
Affinity issue looks fixed and I don't see the "Warning: the number of SMP threads" message:
Charm++> cpuaffinity PE-core map : 0-83:4,88-171:4
Charm++> Running on 4 hosts (1 sockets x 7 cores x 1 PUs = 7-way SMP)

**Comment by jcphill** [comment]
When I provide an invalid pemap on Summit I get a segfault with no error message.

**Comment by evan-charmworks** [comment]
Do you have a backtrace for that?

**Comment by jcphill** [comment]
Nope. Trying to figure out why Summit isn't printing backtraces.

**Comment by jcphill** [comment]
PAMI doesn't print backtraces but MPI does:
Charm++> cpu affinity enabled. 
Charm++> cpuaffinity PE-core map : 0-41
Charm++> set PE 6 on node 0 to core #6
Charm++> set PE 3 on node 0 to core #3
Charm++> set PE 1 on node 0 to core #1
Charm++> set PE 4 on node 0 to core #4
Charm++> set PE 5 on node 0 to core #5
Charm++> set PE 2 on node 0 to core #2
Charm++> set PE 9 on node 1 to core #9
Charm++> set PE 12 on node 1 to core #12
Charm++> set PE 10 on node 1 to core #10
Charm++> set PE 7 on node 1 to core #7
Charm++> set PE 13 on node 1 to core #13
Charm++> set PE 11 on node 1 to core #11
Charm++> set PE 8 on node 1 to core #8
Charm++> set PE 0 on node 0 to core #0
[a23n04:66476] *** Process received signal ***
[a23n04:66476] Signal: Segmentation fault (11)
[a23n04:66476] Signal code: Address not mapped (1)
[a23n04:66476] Failing at address: 0xa0
[a23n04:66476] [ 0] [0x2000000504d8]
[a23n04:66476] [ 1] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2(CmiSetCPUAffinity+0x60)[0x1122c440]
[a23n04:66476] [ 2] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2(CmiInitCPUAffinity+0x47c)[0x1122d57c]
[a23n04:66476] [ 3] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2(_Z10_initCharmiPPc+0x2390)[0x110374b0]
[a23n04:66476] [ 4] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2(_Z10slave_initiPPc+0x1e0)[0x10369320]
[a23n04:66476] [ 5] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2[0x11202f8c]
[a23n04:66476] [ 6] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2[0x11205b10]
[a23n04:66476] [ 7] /lib64/libpthread.so.0(+0x8b94)[0x2000001f8b94]
[a23n04:66476] [ 8] /lib64/libc.so.6(clone+0xe4)[0x2000088a85f4]
[a23n04:66476] *** End of error message ***
[a23n04:66475] *** Process received signal ***
[a23n04:66475] Signal: Segmentation fault (11)
[a23n04:66475] Signal code: Address not mapped (1)
[a23n04:66475] Failing at address: 0xa0
[a23n04:66475] [ 0] [0x2000000504d8]
[a23n04:66475] [ 1] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2(CmiSetCPUAffinity+0x60)[0x1122c440]
[a23n04:66475] [ 2] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2(CmiInitCPUAffinity+0x47c)[0x1122d57c]
[a23n04:66475] [ 3] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2(_Z10_initCharmiPPc+0x2390)[0x110374b0]
[a23n04:66475] [ 4] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2(_Z10slave_initiPPc+0x1e0)[0x10369320]
[a23n04:66475] [ 5] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2[0x11202f8c]
[a23n04:66475] [ 6] namd2/Linux-POWER-xlC.cudamemoptmpi/namd2[0x11205b10]
[a23n04:66475] [ 7] /lib64/libpthread.so.0(+0x8b94)[0x2000001f8b94]
[a23n04:66475] [ 8] /lib64/libc.so.6(clone+0xe4)[0x2000088a85f4]
[a23n04:66475] *** End of error message ***
ERROR:  One or more process (first noticed rank 1) terminated with signal 11

**Comment by jcphill** [comment]
I can reproduce the segfault on multicore-linux64 with simplearrayhello:
(gdb) run +p4 +showcpuaffinity +pemap 13-23
Starting program: /tmp/jun2019/charm/multicore-linux64/tests/charm++/simplearrayhello/hello +p4 +showcpuaffinity +pemap 13-23
Missing separate debuginfos, use: debuginfo-install glibc-2.12-1.149.el6_6.5.x86_64
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Charm++: standalone mode (not using charmrun)
Charm++> Running in Multicore mode: 4 threads (PEs)
[New Thread 0x7ffff7fb7700 (LWP 15073)]
[New Thread 0x7fffeffff700 (LWP 15074)]
[New Thread 0x7ffff75b6700 (LWP 15075)]
Converse/Charm++ Commit ID: v6.9.0-399-g12e54f9
CharmLB> Load balancer assumes all CPUs are same.
Charm++> cpu affinity enabled. 
Charm++> cpuaffinity PE-core map : 13-23
Charm++> set PE 2 on node 0 to core #15
Charm++> set PE 3 on node 0 to core #16
Charm++> set PE 0 on node 0 to core #13
Charm++> set PE 1 on node 0 to core #14
HWLOC> [0] Thread 0x7ffff7fbace0 bound to cpuset: 0x00002000
HWLOC> [2] Thread 0x7fffeffff700 bound to cpuset: 0x00008000
HWLOC> [1] Thread 0x7ffff7fb7700 bound to cpuset: 0x00004000

Thread 4 "hello" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff75b6700 (LWP 15075)]
0x00000000006d24e2 in CmiSetCPUAffinity ()
Missing separate debuginfos, use: debuginfo-install libgcc-4.4.7-11.el6.x86_64 libstdc++-4.4.7-11.el6.x86_64 zlib-1.2.3-29.el6.x86_64
(gdb) where
#0  0x00000000006d24e2 in CmiSetCPUAffinity ()
#1  0x00000000006d3ea7 in CmiInitCPUAffinity ()
#2  0x000000000059f3d1 in _initCharm(int, char**) ()
#3  0x00000000006b0763 in ConverseRunPE(int) ()
#4  0x00000000006ada84 in call_startfn(void*) ()
#5  0x0000003e2ae079d1 in start_thread () from /lib64/libpthread.so.0
#6  0x0000003e2aae88fd in clone () from /lib64/libc.so.6

**Comment by evan-charmworks** [comment]
@jcphill Thanks for the instructions to reproduce with multicore. I've pushed a commit that should fix the null pointer dereference.

**Comment by jcphill** [comment]
Thanks.  Testing on Summit...

**Comment by jcphill** [comment]
Yes, error messages are printed properly now.

**Comment by evan-charmworks** [comment]
I'll fix it when squashing.

**Pull request merged**